src/test/optimizer/strategy/semi_join_reduction_rule_test.cpp

```diff
@@ -0,0 +1,190 @@
+#include "optimizer/strategy/strategy_base_test.hpp"
+
+#include "optimizer/strategy/semi_join_reduction_rule.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+class SemiJoinReductionRuleTest : public StrategyBaseTest {
+ protected:
+  void SetUp() override {
+    {
+      const auto histogram_column_a = GenericHistogram<int32_t>::with_single_bin(1, 50, 40, 40);
+      const auto histogram_column_b = GenericHistogram<int32_t>::with_single_bin(10, 15, 40, 5);
+      _node_a = create_mock_node_with_statistics({{DataType::Int, "a"}, {DataType::Int, "b"}}, 40,
+                                                 {histogram_column_a, histogram_column_b});
+      _a_a = _node_a->get_column("a");
+      _a_b = _node_a->get_column("b");
+    }
+
+    {
+      const auto histogram_column_a = GenericHistogram<int32_t>::with_single_bin(10, 20, 10, 10);
+      const auto histogram_column_b = GenericHistogram<int32_t>::with_single_bin(40, 60, 10, 5);
+      _node_b = create_mock_node_with_statistics({{DataType::Int, "a"}, {DataType::Int, "b"}}, 10,
+                                                 {histogram_column_a, histogram_column_b});
+      _b_a = _node_b->get_column("a");
+      _b_b = _node_b->get_column("b");
+    }
+
+    {
+      const auto histogram_column_a = GenericHistogram<int32_t>::with_single_bin(10, 15, 40, 5);
+      _node_c = create_mock_node_with_statistics({{DataType::Int, "a"}}, 40, {histogram_column_a});
+      _c_a = _node_c->get_column("a");
+    }
+  }
+
+  std::shared_ptr<MockNode> _node_a, _node_b, _node_c;
+  LQPColumnReference _a_a, _a_b, _b_a, _b_b, _c_a;
+  std::shared_ptr<SemiJoinReductionRule> _rule{std::make_shared<SemiJoinReductionRule>()};
+};
+
+TEST_F(SemiJoinReductionRuleTest, CreateSimpleReduction) {
+  // The _a_a side of the inner join has values from 1-50, the _b_a side has values from 10-20. Based on that
+  // selectivity, a semi join reduction should be created.
+
+  // clang-format off
+  const auto input_lqp = JoinNode::make(JoinMode::Inner, equals_(_a_a, _b_a),
+    _node_a,
+    _node_b);
+
+  const auto expected_lqp = JoinNode::make(JoinMode::Inner, equals_(_a_a, _b_a),
```

Comment:
Am I right that there is no specific test that the `try_deeper` method really hangs the entry point to the "correct" position?

Reply:
SemiJoinReductionRuleTest.TraverseRightInput